vignettes/task_plans.Rmd

```diff
@@ -0,0 +1,353 @@
+
+One drawback of the `add_complete` system and corresponding .ind files is that the speed shortcut only works if the presence of these .ind files is a fairly accurate description of which tasks are up to date. If you build a complete task plan once, change an input, and intend to rebuild, then it's strongly recommended that you delete the .ind files before calling `loop_tasks` again. If you don't, nothing will break, but your build will be slower and will not include automatic retries.
+
+```{r plan}
+task_plan_1 <- create_task_plan(
+  task_config$id,
+  list(step1, step2, step3),
+  final_steps='report',
+  ind_dir='states/log')
+```
+
+You'll need the `task_plan` R object for two things: to create a remake .yml makefile and to run `loop_tasks`.
+
+
+#### Creating a task makefile
+
+Here's how to create a task makefile from a task plan:
+
+```{r makefile, results='markup'}
+create_task_makefile(
+  task_plan=task_plan_1,
+  makefile='task_plan_1.yml',
+  include=c(),
+  packages=c('ggplot2'),
+  sources=c('script.R'),
+  file_extensions=c('ind'),
+  ind_complete=TRUE)
+```
+
+Here are the contents of task_plan_1.yml:
+
+```{r read_makefile, results='markup', comment='', echo=FALSE}
+cat(readr::read_lines('task_plan_1.yml'), sep='\n')
+```
+
+Note that at the end of each state's task list is a final task, e.g., `states/log/WI.ind`, that we didn't directly specify in the task_plan. These recipes for the `complete` step were generated in `create_task_plan` as described above. These .ind files representing the completeness of each task are also used as the dependencies of the final target, `task_plan_1`, which describes all targets that must be complete to consider the whole task batch complete.
+
+
+#### Executing a task plan/makefile
+
+You can use `loop_tasks` with the `task_plan` and the `task_makefile` to run all the steps in all the tasks:
+
+```{r delete1, echo=FALSE}
+unlink('.remake')
+unlink('build')
+```
+
+```{r loop1a}
+loop_tasks(
+  task_plan=task_plan_1,
+  task_makefile='task_plan_1.yml')
+```
+
+The benefits of using `loop_tasks` to run all the remake targets in task_plan_1.yml are:
+
+1. baked-in retries during each execution of `loop_tasks`
+2. quicker restarts if `loop_tasks` gets interrupted partway through (e.g., if your computer runs out of battery or crashes)
+3. the ability to easily run subsets of tasks and/or steps while you're developing the analysis (see below)
+
+To use the baked-in retries effectively, simply ensure that the functions called in your task-steps throw errors (typically with `stop()`) when things go wrong. No need to build retries into the functions themselves because `loop_tasks` will take care of that for you. And if your function isn't the sort that would benefit from retries - e.g., HTTP requests need retries, but a deterministic algorithm should just give up if it breaks - you can set `num_tries=1` in `loop_tasks`.
+
```

Comment:
It might be good to add a sentence or paragraph here about how we've learned that `loop_tasks` can be waaaay slower than a simple `scmake` for large task makefiles. I wish it just weren't slower, but I don't think we're going to fix that anytime soon, and in the meantime, it pays to know that you have both options. The  command for running this task plan with `scmake` would be
```r
scmake('task_plan_1', 'task_plan_1.yml')
```
and a remake recipe for it in the parent makefile might be
```yml
  states/log/task_plan_1.ind:
    command: scmake('task_plan_1', 'task_plan_1.yml', force=TRUE)
```
where the `force=TRUE` within the command would mean that even though the file `states/log/task_plan_1.ind` only gets built once on its own, you could call `scmake('states/log/task_plan_1.ind', force=TRUE)` at the prompt and because the recipe command also has force=TRUE, the indicator file would actually get rebuilt. (This still doesn't rebuild the individual task-steps, but it does avoid one sometimes-surprising behavior (failure to rebuild when you call force=TRUE at the prompt but not in the parent makefile).

Reply:
...And maybe that last idea, the addition of `force=TRUE`, would be good to save for the next section.